File: DiKErnel/Src/DiKErnel.FunctionLibrary/Public/GrassRevetmentOvertopping/GrassRevetmentOvertoppingFunctions.cpp
// Copyright (C) Stichting Deltares 2022. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include "GrassRevetmentOvertoppingFunctions.h"

#include <algorithm>
#include <cmath>
#include <functional>

#include "Geometry.h"
#include "GrassRevetmentFunctions.h"
#include "Input.h"
#include "Load.h"
#include "OvertoppingAdapter.h"

namespace DiKErnel::FunctionLibrary
{
    using namespace std;
    using namespace External::Overtopping;

    double GrassRevetmentOvertoppingFunctions::RepresentativeWaveRunup2P(
        const GrassRevetmentOvertoppingRepresentative2PInput& input)
    {
        Load load
        {
            ._waterLevel = input._waterLevel,
            ._height = input._waveHeightHm0,
            ._period = input._wavePeriodTm10,
            ._direction = input._waveDirection
        };

        Geometry geometry
        {
            ._normal = 0.0,
            ._nPoints = static_cast<int>(input._xValuesProfile.size()),
            ._xCoords = input._xValuesProfile.data(),
            ._yCoords = input._zValuesProfile.data(),
            ._roughness = input._roughnessCoefficients.data()
        };

        Input overtoppingInput{
            ._factorDeterminationQnFn = 1.0,
            ._factorDeterminationQbFb = 1.0,
            ._mz2 = 1.0,
            ._fshallow = 1.0,
            ._computedOvertopping = 1.0,
            ._criticalOvertopping = 1.0,
            ._relaxationfactor = 1.0,
            ._reductionFactorForeshore = 0.5
        };

        Result result{};

        bool success = false;
        const auto messageBuffer = make_unique<string>();
        messageBuffer->reserve(255);

        OvertoppingAdapter::CalculateQo(load, geometry, overtoppingInput, &result, messageBuffer.get(), &success, input._dikeHeight);

        return result._z2;
    }

    double GrassRevetmentOvertoppingFunctions::CumulativeOverload(
        const GrassRevetmentOvertoppingCumulativeOverloadInput& input)
    {
        const function getFrontVelocity = [&input](
            const double waveRunup)
        {
            return FrontVelocity(waveRunup, input._verticalDistanceWaterLevelElevation, input._accelerationAlphaA, input._frontVelocityCwo,
                                 input._gravitationalAcceleration);
        };

        return GrassRevetmentFunctions::CumulativeOverload(input, getFrontVelocity);
    }

    double GrassRevetmentOvertoppingFunctions::FrontVelocity(
        const double waveRunup,
        const double verticalDistanceWaterLevelElevation,
        const double accelerationAlphaA,
        const double frontVelocityCwo,
        const double gravitationalAcceleration)
    {
        return frontVelocityCwo * accelerationAlphaA * sqrt(gravitationalAcceleration * max(0.0, waveRunup - verticalDistanceWaterLevelElevation));
    }
}

File: Src/DiKErnel.FunctionLibrary/GrassRevetmentOvertopping/GrassRevetmentOvertoppingFunctions.cs
// // Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
// //
// // This file is part of DiKErnel.
// //
// // DiKErnel is free software: you can redistribute it and/or modify it under the terms of the
// // GNU Lesser General Public License as published by the Free Software Foundation, either
// // version 3 of the License, or (at your option) any later version.
// //
// // This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// // without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// // See the GNU Lesser General Public License for more details.
// //
// // You should have received a copy of the GNU Lesser General Public License along with this
// // program. If not, see <http://www.gnu.org/licenses/>.
// //
// // All names, logos, and references to "Deltares" are registered trademarks of Stichting
// // Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System;
using DiKErnel.External.Overtopping;
using DiKErnel.FunctionLibrary.GrassRevetment;

namespace DiKErnel.FunctionLibrary.GrassRevetmentOvertopping
{
    /// <summary>
    /// Class that holds all grass revetment overtopping specific calculation routines.
    /// </summary>
    public static class GrassRevetmentOvertoppingFunctions
    {
        /// <summary>
        /// Calculates the representative wave run-up (2 percent).
        /// </summary>
        /// <param name="input">The input to use for the calculation.</param>
        /// <returns>The representative wave run-up (2 percent) [m].</returns>
        public static double RepresentativeWaveRunup2P(GrassRevetmentOvertoppingRepresentative2PInput input)
        {
            return OvertoppingAdapter.CalculateZ2(input.WaterLevel, input.WaveHeightHm0, input.WavePeriodTm10, input.WaveDirection,
                                                  input.XValuesProfile, input.ZValuesProfile, input.RoughnessCoefficients,
                                                  input.DikeHeight);
        }

        /// <summary>
        /// Calculates the cumulative overload.
        /// </summary>
        /// <param name="input">The input to use for the calculation.</param>
        /// <returns>The cumulative overload [m^2/s^2].</returns>
        public static double CumulativeOverload(GrassRevetmentOvertoppingCumulativeOverloadInput input)
        {
            Func<double, double> getFrontVelocityFunc = waveRunup => FrontVelocity(waveRunup,
                                                                                   input.VerticalDistanceWaterLevelElevation,
                                                                                   input.AccelerationAlphaA,
                                                                                   input.FrontVelocityCwo,
                                                                                   input.GravitationalAcceleration);

            return GrassRevetmentFunctions.CumulativeOverload(input, getFrontVelocityFunc);
        }

        private static double FrontVelocity(double waveRunup, double verticalDistanceWaterLevelElevation,
                                            double accelerationAlphaA, double frontVelocityCwo, double gravitationalAcceleration)
        {
            return frontVelocityCwo * accelerationAlphaA
                                    * Math.Sqrt(gravitationalAcceleration
                                                * Math.Max(0.0, waveRunup - verticalDistanceWaterLevelElevation));
        }
    }
}